### PR TITLE
Make `GrpcStreamBroadcaster` compatible with both `grpcio` and `grpclib`

### PR DESCRIPTION
**Pull request by llucax**
With this change `GrpcStreamBroadcaster` can be used with both `grpcio` and `grpclib` implementations of gRPC.

This is achieved by using `AsyncIterator` instead of `grpc.aio.UnaryStreamCall` in the `stream_method` parameter of `GrpcStreamBroadcaster` to make it more generic, and catching exceptions from both `grpcio` and `grpclib` in the `async for` loop.

To deal with any of the libraries being installed, a new module is added to handles exceptions from both libraries, by conditionally importing the exception and provide a dummy one if the library is not installed.

Now both `grpcio` and `grpclib` are optional, and can be installed using `frequenz-client-base[grpcio]` or `frequenz-client-base[grpclib]` respectively.

We also add some tests for `GrpcStreamBroadcaster`.

### CODE REVIEW

**Comment by llucax** [comment]
Based on #48.

File: tests/streaming/test_grpc_stream_broadcaster.py
Comment (daniel-zullo-frequenz) [inline tests/streaming/test_grpc_stream_broadcaster.py]: ~Maybe `grpc_error` or just `error` instead?~
```suggestion
async def test_streaming_grpc(  # pylint: disable=too-many-arguments
```

File: src/frequenz/client/base/_grpchacks.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/client/base/_grpchacks.py:4]: I think it is fine to mention this is a hack in the documentation. Though I personally don't like to have the word `hacks` in the module name. What do you think about renaming the module from `_grpchacks.py` to something around `_grpc_utils.py`? (I didn't suggest `_grpc_base_exceptions.py` as I understood it'd be too specific to the current state and not flexible for extending it)

File: tests/streaming/test_grpc_stream_broadcaster.py
Comment (daniel-zullo-frequenz) [inline tests/streaming/test_grpc_stream_broadcaster.py]: This kind of contradict the name of the test. IIUC he test itself checks for both, success and error

File: src/frequenz/client/base/_grpchacks.py
Comment (llucax) [reply src/frequenz/client/base/_grpchacks.py:4]: I intentionally used  `hacks` to make it look dirty. I considered `utils` but it is too broad, and it could invite people to add random stuff there. I picked hacks because it entitles that it is not something desirable to have, and that hopefully it is temporary. If you had a bad reaction to it, I think my naming was successful :laughing:

File: tests/streaming/test_grpc_stream_broadcaster.py
Comment (llucax) [reply tests/streaming/test_grpc_stream_broadcaster.py]: Will change to  `test_streaming_error`.

File: tests/streaming/test_grpc_stream_broadcaster.py
Comment (llucax) [reply tests/streaming/test_grpc_stream_broadcaster.py]: Copy&pasta :spaghetti:.  Will fix.

**Comment by llucax** [comment]
Updated. Will fix up before merging.

File: src/frequenz/client/base/_grpchacks.py
Comment (daniel-zullo-frequenz) [reply src/frequenz/client/base/_grpchacks.py:4]: It did trigger some noise in my mind :D 
Hopefully it won't stay forever :crossed_fingers:

**Review by daniel-zullo-frequenz — COMMENTED** [review]
LGTM, it only needs fixing up as stated...

**Comment by llucax** [comment]
Updated.

**Comment by llucax** [comment]
Enabled auto-merge.

**Review by daniel-zullo-frequenz — DISMISSED** [review]

**Comment by llucax** [comment]
The good news is the merge queue checks now works correctly, the bad news is I forgot 2 fixes need to be applied to fix the CI, so this PR depends now on this PR to be merged:

* https://github.com/frequenz-floss/frequenz-client-base-python/pull/50

**Review by daniel-zullo-frequenz — DISMISSED** [review]

**Comment by llucax** [comment]
Rebased, it needs a new approval.

**Review by daniel-zullo-frequenz — APPROVED** [review]

**Pull request merged**
